azkaban-common/src/main/java/azkaban/executor/ExecutionControllerUtils.java

```diff
@@ -175,30 +177,38 @@ public static void finalizeFlow(final EventHandler eventHandler,
    * @param originalStatus
    */
   public static void restartFlow(final ExecutableFlow flow, final Status originalStatus) {
-    if (!RESTARTABLE_STATUSES.contains(originalStatus)) {
-      return;
+    if (!RESTARTABLE_NON_TERMINATED_STATUSES.contains(originalStatus)
+        && !RESTARTABLE_TERMINATED_STATUSES.contains(originalStatus)) {
+        return;
     }
     final ExecutionOptions options = flow.getExecutionOptions();
     // flow can only be retried once
     if (options == null || options.isExecutionRetried()) {
       return;
     }
```

Comment:
It looks like this code block will prevent us from honoring the property `flow.max.retries`. How are handling that? If it's going to be addressed separately could you please add a TODO?

Reply:
Good point, as for now it's not using `flow.max.retries` for now and will add a TODO